<a href="https://colab.research.google.com/github/VOL-EM2024/mo/blob/main/%22%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_1_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл
hghg
**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv("/content/global-data-on-sustainable-energy (1).csv")
df

Entity  Year  Access to electricity (% of population)  \
0     Afghanistan  2000                                 1.613591   
1     Afghanistan  2001                                 4.074574   
2     Afghanistan  2002                                 9.409158   
3     Afghanistan  2003                                14.738506   
4     Afghanistan  2004                                20.064968   
...           ...   ...                                      ...   
3644     Zimbabwe  2016                                42.561730   
3645     Zimbabwe  2017                                44.178635   
3646     Zimbabwe  2018                                45.572647   
3647     Zimbabwe  2019                                46.781475   
3648     Zimbabwe  2020                                52.747670   

      Access to clean fuels for cooking  \
0                                   6.2   
1                                   7.2   
2                                   8.2   
3                                   9.5   
4                                  10.9   
...                                 ...   
3644                               29.8   
3645                               29.8   
3646                               29.9   
3647                               30.1   
3648                               30.4   

      Renewable-electricity-generating-capacity-per-capita  \
0                                                  9.22      
1                                                  8.86      
2                                                  8.47      
3                                                  8.09      
4                                                  7.75      
...                                                 ...      
3644                                              62.88      
3645                                              62.33      
3646                                              82.53      
3647                                              81.40      
3648                                              80.61      

      Financial flows to developing countries (US $)  \
0                                            20000.0   
1                                           130000.0   
2                                          3950000.0   
3                                         25970000.0   
4                                                NaN   
...                                              ...   
3644                                         30000.0   
3645                                       5570000.0   
3646                                         10000.0   
3647                                        250000.0   
3648                                         30000.0   

      Renewable energy share in the total final energy consumption (%)  \
0                                                 44.99                  
1                                                 45.60                  
2                                                 37.83                  
3                                                 36.66                  
4                                                 44.24                  
...                                                 ...                  
3644                                              81.90                  
3645                                              82.46                  
3646                                              80.23                  
3647                                              81.50                  
3648                                              81.90                  

      Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
0                                    0.16                             0.0   
1                                    0.09                             0.0   
2                                    0.13                             0.0   
3                                    0.31                     

##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [ ]:
#Колонки
df.columns

Index(['Entity', 'Year', 'Access to electricity (% of population)',
       'Access to clean fuels for cooking',
       'Renewable-electricity-generating-capacity-per-capita',
       'Financial flows to developing countries (US $)',
       'Renewable energy share in the total final energy consumption (%)',
       'Electricity from fossil fuels (TWh)', 'Electricity from nuclear (TWh)',
       'Electricity from renewables (TWh)',
       'Low-carbon electricity (% electricity)',
       'Primary energy consumption per capita (kWh/person)',
       'Energy intensity level of primary energy (MJ/$2017 PPP GDP)',
       'Value_co2_emissions_kt_by_country',
       'Renewables (% equivalent primary energy)', 'gdp_growth',
       'gdp_per_capita', 'Density\n(P/Km2)', 'Land Area(Km2)', 'Latitude',
       'Longitude'],
      dtype='object')

In [ ]:
#Первые 5 значений
df.head(5)

Entity  Year  Access to electricity (% of population)  \
0  Afghanistan  2000                                 1.613591   
1  Afghanistan  2001                                 4.074574   
2  Afghanistan  2002                                 9.409158   
3  Afghanistan  2003                                14.738506   
4  Afghanistan  2004                                20.064968   

   Access to clean fuels for cooking  \
0                                6.2   
1                                7.2   
2                                8.2   
3                                9.5   
4                               10.9   

   Renewable-electricity-generating-capacity-per-capita  \
0                                               9.22      
1                                               8.86      
2                                               8.47      
3                                               8.09      
4                                               7.75      

   Financial flows to developing countries (US $)  \
0                                         20000.0   
1                                        130000.0   
2                                       3950000.0   
3                                      25970000.0   
4                                             NaN   

   Renewable energy share in the total final energy consumption (%)  \
0                                              44.99                  
1                                              45.60                  
2                                              37.83                  
3                                              36.66                  
4                                              44.24                  

   Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
0                                 0.16                             0.0   
1                                 0.09                             0.0   
2                                 0.13                             0.0   
3                                 0.31                             0.0   
4                                 0.33                             0.0   

   Electricity from renewables (TWh)  ...  \
0                               0.31  ...   
1                               0.50  ...   
2                               0.56  ...   
3                               0.63  ...   
4                               0.56  ...   

   Primary energy consumption per capita (kWh/person)  \
0                                          302.59482    
1                                          236.89185    
2                                          210.86215    
3                                          229.96822    
4                                          204.23125    

   Energy intensity level of primary energy (MJ/$2017 PPP GDP)  \
0                                               1.64             
1                                               1.74             
2                                               1.40             
3                                               1.40             
4                                               1.20             

   Value_co2_emissions_kt_by_country  \
0                         760.000000   
1                         730.000000   
2                        1029.999971   
3                        1220.000029   
4                        1029.999971   

   Renewables (% equivalent primary energy)  gdp_growth  gdp_per_capita  \
0                                       NaN         NaN             NaN   
1                                       NaN         NaN             NaN   
2                                       NaN         NaN      179.426579   
3                                       NaN    8.832278      190.683814   
4                                       NaN    1.414118      211.382074   

   Density\n(P/Km2) Land Area(Km2)  Latitude  Longitude  
0                60       652230.0  33.93911  67.709953  
1          

In [ ]:
#Последние 5 значений
df.tail(5)

Entity  Year  Access to electricity (% of population)  \
3644  Zimbabwe  2016                                42.561730   
3645  Zimbabwe  2017                                44.178635   
3646  Zimbabwe  2018                                45.572647   
3647  Zimbabwe  2019                                46.781475   
3648  Zimbabwe  2020                                52.747670   

      Access to clean fuels for cooking  \
3644                               29.8   
3645                               29.8   
3646                               29.9   
3647                               30.1   
3648                               30.4   

      Renewable-electricity-generating-capacity-per-capita  \
3644                                              62.88      
3645                                              62.33      
3646                                              82.53      
3647                                              81.40      
3648                                              80.61      

      Financial flows to developing countries (US $)  \
3644                                         30000.0   
3645                                       5570000.0   
3646                                         10000.0   
3647                                        250000.0   
3648                                         30000.0   

      Renewable energy share in the total final energy consumption (%)  \
3644                                              81.90                  
3645                                              82.46                  
3646                                              80.23                  
3647                                              81.50                  
3648                                              81.90                  

      Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
3644                                 3.50                             0.0   
3645                                 3.05                             0.0   
3646                                 3.73                             0.0   
3647                                 3.66                             0.0   
3648                                 3.40                             0.0   

      Electricity from renewables (TWh)  ...  \
3644                               3.32  ...   
3645                               4.30  ...   
3646                               5.46  ...   
3647                               4.58  ...   
3648                               4.19  ...   

      Primary energy consumption per capita (kWh/person)  \
3644                                          3227.6802    
3645                                          3068.0115    
3646                                          3441.9858    
3647                                          3003.6553    
3648                                          2680.1318    

      Energy intensity level of primary energy (MJ/$2017 PPP GDP)  \
3644                                              10.00             
3645                                               9.51             
3646                                               9.83             
3647                                              10.47             
3648                                              10.00             

      Value_co2_emissions_kt_by_country  \
3644                        11020.00046   
3645                        10340.00015   
3646                        12380.00011   
3647                        11760.00023   
3648                                NaN   

      Renewables (% equivalent primary energy)  gdp_growth  gdp_per_capita  \
3644                                       NaN    0.755869     1464.588957   
3645                                       NaN    4.709492     1235.189032   
3646                                       NaN    4.824211     1254.642265   
3647                                       NaN   -6.144236     1316.740657   
3648                      

In [ ]:
#Статистику
df.describe()

Year  Access to electricity (% of population)  \
count  3649.000000                              3639.000000   
mean   2010.038367                                78.933702   
std       6.054228                                30.275541   
min    2000.000000                                 1.252269   
25%    2005.000000                                59.800890   
50%    2010.000000                                98.361570   
75%    2015.000000                               100.000000   
max    2020.000000                               100.000000   

       Access to clean fuels for cooking  \
count                        3480.000000   
mean                           63.255287   
std                            39.043658   
min                             0.000000   
25%                            23.175000   
50%                            83.150000   
75%                           100.000000   
max                           100.000000   

       Renewable-electricity-generating-capacity-per-capita  \
count                                        2718.000000      
mean                                          113.137498      
std                                           244.167256      
min                                             0.000000      
25%                                             3.540000      
50%                                            32.910000      
75%                                           112.210000      
max                                          3060.190000      

       Financial flows to developing countries (US $)  \
count                                    1.560000e+03   
mean                                     9.422400e+07   
std                                      2.981544e+08   
min                                      0.000000e+00   
25%                                      2.600000e+05   
50%                                      5.665000e+06   
75%                                      5.534750e+07   
max                                      5.202310e+09   

       Renewable energy share in the total final energy consumption (%)  \
count                                        3455.000000                  
mean                                           32.638165                  
std                                            29.894901                  
min                                             0.000000                  
25%                                             6.515000                  
50%                                            23.300000                  
75%                                            55.245000                  
max                                            96.040000                  

       Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
count                          3628.000000                     3523.000000   
mean                             70.365003                       13.450190   
std                             348.051866                       73.006623   
min                               0.000000                        0.000000   
25%                               0.290000                        0.000000   
50%                               2.970000                        0.000000   
75%                              26.837500                        0.000000   
max                            5184.130000                      809.410000   

       Electricity from renewables (TWh)  \
count                        3628.000000   
mean                           23.968010   
std                           104.431085   
min                             0.000000   
25%                             0.040000   
50%                             1.470000   
75%                             9.600000   
max                          2184.940000   

       Low-carbon electricity (% electricity)  \
count                             3607.000000   
mean                                36.801182   
std                                 34.31488

**Ответьте на следующие вопросы:**


1.   Какие средние значения в каждой числовой колонке вашего датасета?
2.   Минимальные и максимальные значения в каждой числовой колонке вашего датасета





1.   **Ответ 1. 2010.038367	78.933702	63.255287	113.137498	9.422400e+07	32.638165	70.365003	13.450190	23.968010	36.801182	25743.981745	5.307345	1.598665e+05	11.986707	3.441610	13283.774348	6.332135e+05	18.246388	14.822695**
2.   **Ответ 2. min - 2000.000000	1.252269	0.000000	0.000000	0.000000e+00	0.000000	0.000000	0.000000	0.000000	0.000000	0.000000	0.110000	1.000000e+01	0.000000	-62.075920	111.927225	2.100000e+01	-40.900557	-175.198242
max -  2020.000000	100.000000	100.000000	3060.190000	5.202310e+09	96.040000	5184.130000	809.410000	2184.940000	100.000010	262585.700000	32.570000	1.070722e+07	86.836586	123.139555	123514.196700	9.984670e+06	64.963051	178.065032**



##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [ ]:
df.columns

Index(['Entity', 'Year', 'Access to electricity (% of population)',
       'Access to clean fuels for cooking',
       'Renewable-electricity-generating-capacity-per-capita',
       'Financial flows to developing countries (US $)',
       'Renewable energy share in the total final energy consumption (%)',
       'Electricity from fossil fuels (TWh)', 'Electricity from nuclear (TWh)',
       'Electricity from renewables (TWh)',
       'Low-carbon electricity (% electricity)',
       'Primary energy consumption per capita (kWh/person)',
       'Energy intensity level of primary energy (MJ/$2017 PPP GDP)',
       'Value_co2_emissions_kt_by_country',
       'Renewables (% equivalent primary energy)', 'gdp_growth',
       'gdp_per_capita', 'Density\n(P/Km2)', 'Land Area(Km2)', 'Latitude',
       'Longitude'],
      dtype='object')

In [ ]:
df["Financial flows to developing countries (US $)"].fillna(0).astype("int64")

0          20000
1         130000
2        3950000
3       25970000
4              0
          ...   
3644       30000
3645     5570000
3646       10000
3647      250000
3648       30000
Name: Financial flows to developing countries (US $), Length: 3649, dtype: int64

In [ ]:

df.info()
#Вывести типы колонок тут

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3649 entries, 0 to 3648
Data columns (total 21 columns):
 #   Column                                                            Non-Null Count  Dtype  
---  ------                                                            --------------  -----  
 0   Entity                                                            3649 non-null   object 
 1   Year                                                              3649 non-null   int64  
 2   Access to electricity (% of population)                           3639 non-null   float64
 3   Access to clean fuels for cooking                                 3480 non-null   float64
 4   Renewable-electricity-generating-capacity-per-capita              2718 non-null   float64
 5   Financial flows to developing countries (US $)                    1560 non-null   float64
 6   Renewable energy share in the total final energy consumption (%)  3455 non-null   float64
 7   Electricity from fossil fuels (TW

Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?

##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [ ]:
fil1_df=df[['Year','Latitude']]
fil1_df

Year   Latitude
0     2000  33.939110
1     2001  33.939110
2     2002  33.939110
3     2003  33.939110
4     2004  33.939110
...    ...        ...
3644  2016 -19.015438
3645  2017 -19.015438
3646  2018 -19.015438
3647  2019 -19.015438
3648  2020 -19.015438

[3649 rows x 2 columns]

In [4]:
fil2_df=df[['Latitude','Longitude']]
fil2_df

Latitude  Longitude
0     33.939110  67.709953
1     33.939110  67.709953
2     33.939110  67.709953
3     33.939110  67.709953
4     33.939110  67.709953
...         ...        ...
3644 -19.015438  29.154857
3645 -19.015438  29.154857
3646 -19.015438  29.154857
3647 -19.015438  29.154857
3648 -19.015438  29.154857

[3649 rows x 2 columns]

In [5]:
fil3_df=df[['Electricity from fossil fuels (TWh)','Electricity from nuclear (TWh)']]
fil3_df

Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)
0                                    0.16                             0.0
1                                    0.09                             0.0
2                                    0.13                             0.0
3                                    0.31                             0.0
4                                    0.33                             0.0
...                                   ...                             ...
3644                                 3.50                             0.0
3645                                 3.05                             0.0
3646                                 3.73                             0.0
3647                                 3.66                             0.0
3648                                 3.40                             0.0

[3649 rows x 2 columns]

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [6]:
gr=df.groupby('Year').mean()
gr

<ipython-input-6-07660c2d883a>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  gr=df.groupby('Year').mean()


Access to electricity (% of population)  \
Year                                            
2000                                73.121292   
2001                                73.646500   
2002                                73.992437   
2003                                74.564801   
2004                                75.147966   
2005                                75.560816   
2006                                76.492228   
2007                                76.913793   
2008                                77.440696   
2009                                77.804968   
2010                                78.534954   
2011                                79.403946   
2012                                80.275411   
2013                                80.318206   
2014                                81.048829   
2015                                81.683522   
2016                                82.777890   
2017                                83.562116   
2018                                84.328949   
2019                                84.760534   
2020                                85.460078   

      Access to clean fuels for cooking  \
Year                                      
2000                          58.063110   
2001                          58.628659   
2002                          59.192683   
2003                          59.779878   
2004                          60.332317   
2005                          60.902744   
2006                          61.474085   
2007                          62.052108   
2008                          62.603313   
2009                          63.123494   
2010                          63.642470   
2011                          64.154518   
2012                          64.621988   
2013                          64.689820   
2014                          65.151796   
2015                          65.563174   
2016                          66.005090   
2017                          66.396707   
2018                          66.795210   
2019                          67.187126   
2020                          67.566168   

      Renewable-electricity-generating-capacity-per-capita  \
Year                                                         
2000                                          88.568231      
2001                                          83.596977      
2002                                          84.579457      
2003                                          84.987132      
2004                                          85.498837      
2005                                          86.348372      
2006                                          86.496279      
2007                                          99.734341      
2008                                         100.574806      
2009                                         105.450155      
2010                                         107.633798      
2011                                         110.515039      
2012                                         114.057984      
2013                                         117.989385      
2014                                         122.649000      
2015                                         129.799538      
2016                                         137.590385      
2017                                         144.333923      
2018                                         149.727154      
2019                                         164.358231      
2020                                         169.797077      

      Financial flows to developing countries (US $)  \
Year                                                   
2000                                    2.572255e+07   
2001                                    2.371608e+07   
2002                                    2.476600e+07   
2003                                    5.896115e+07   
2004                                    2.559237e+07   
2005                                    2.931891e+07   
2006         

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.

In [ ]:
#Агрегация